contracts/protocol/tokenization/NTokenUniswapV3.sol

```diff
@@ -161,31 +192,42 @@ contract NTokenUniswapV3 is NToken {
         ) = INonfungiblePositionManager(_underlyingAsset).positions(tokenId);
 
         // move underlying into this contract
-        IERC20(token0).safeTransferFrom(
-            _msgSender(),
-            address(this),
-            amountAdd0
-        );
-        IERC20(token1).safeTransferFrom(
-            _msgSender(),
-            address(this),
-            amountAdd1
-        );
-
-        uint256 token0Allownance = IERC20(token0).allowance(
-            address(this),
-            _underlyingAsset
-        );
-        if (token0Allownance == 0) {
-            IERC20(token0).safeApprove(_underlyingAsset, 2**256 - 1);
+        address weth = _addressesProvider.getWETH();
+        uint256 txValue = msg.value;
+        bool token0IsETH = (token0 == weth && txValue > 0);
+        bool token1IsETH = (token1 == weth && txValue > 0);
+        if (!token0IsETH) {
+            IERC20(token0).safeTransferFrom(
+                _msgSender(),
+                address(this),
+                amountAdd0
+            );
+        }
+        if (!token1IsETH) {
+            IERC20(token1).safeTransferFrom(
+                _msgSender(),
+                address(this),
+                amountAdd1
+            );
         }
 
-        uint256 token1Allownance = IERC20(token1).allowance(
-            address(this),
-            _underlyingAsset
-        );
-        if (token1Allownance == 0) {
-            IERC20(token1).safeApprove(_underlyingAsset, 2**256 - 1);
+        //avoid stack too deep
+        {
+            uint256 token0Allownance = IERC20(token0).allowance(
+                address(this),
+                _underlyingAsset
+            );
+            if (token0Allownance == 0) {
+                IERC20(token0).safeApprove(_underlyingAsset, 2**256 - 1);
+            }
+
+            uint256 token1Allownance = IERC20(token1).allowance(
+                address(this),
+                _underlyingAsset
+            );
+            if (token1Allownance == 0) {
+                IERC20(token1).safeApprove(_underlyingAsset, 2**256 - 1);
```

Comment:
this is duplicated with token0, can we define a common function?

Reply:
yeah.